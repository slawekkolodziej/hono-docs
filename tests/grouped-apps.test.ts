import { describe, test, expect, beforeAll } from "vitest";
import { execSync } from "child_process";
import fs from "fs";
import path from "path";

// Test files are located in integration-tests directory
const rootDir = path.resolve(import.meta.dirname, "..");
const integrationTestsDir = path.join(rootDir, "integration-tests");
const cliPath = path.join(rootDir, "dist/cli/index.js");

describe.each([
  {
    name: "grouped-simple",
    configPath: "./grouped-simple/hono-docs.config.ts",
    description: "Simple grouped app with .route()",
    expectedRoutes: {
      "/docs": ["get"],
      "/docs/{id}": ["get"],
      "/users": ["get", "post"],
      "/users/{id}": ["get"],
    },
  },
  {
    name: "grouped-basepath",
    configPath: "./grouped-basepath/hono-docs.config.ts",
    description: "Grouped app with basePath",
    expectedRoutes: {
      "/v1/docs": ["get"],
      "/v1/docs/{id}": ["get"],
      "/v1/users": ["get", "post"],
      "/v1/users/{id}": ["get"],
    },
  },
  {
    name: "grouped-complex",
    configPath: "./grouped-complex/hono-docs.config.ts",
    description: "Complex grouped app with mixed composition",
    expectedRoutes: {
      "/health": ["get"],
      "/api/v1/docs": ["get"],
      "/api/v1/docs/{id}": ["get"],
      "/api/v1/users": ["get", "post"],
      "/api/v1/users/{id}": ["get"],
      "/auth/login": ["post"],
      "/auth/logout": ["post"],
    },
  },
  {
    name: "grouped-nested",
    configPath: "./grouped-nested/hono-docs.config.ts",
    description: "Nested grouping",
    expectedRoutes: {
      "/api/docs": ["get"],
      "/api/docs/{id}": ["get"],
      "/api/users": ["get", "post"],
      "/api/users/{id}": ["get"],
    },
  },
])("$name - $description", ({ configPath, expectedRoutes }) => {
  // Run the generation using the local CLI
  const configPathFull = path.join(integrationTestsDir, configPath);
  const testCaseDir = path.dirname(configPathFull);
  const outputPath = path.resolve(testCaseDir, "openapi.json");
  let output: any = null;

  beforeAll(() => {
    execSync(`node ${cliPath} generate --config ${configPathFull}`, {
      stdio: "inherit",
      cwd: testCaseDir,
    });

    output = JSON.parse(fs.readFileSync(outputPath, "utf-8"));
  });

  test("generate OpenAPI spec", () => {
    // Check that output file exists
    expect(fs.existsSync(outputPath)).toBe(true);

    // Basic structure checks
    expect(output).toHaveProperty("openapi");
    expect(output).toHaveProperty("info");
    expect(output).toHaveProperty("paths");

    // Check that all expected routes are present with correct methods
    const paths = output.paths || {};

    expect(paths).toEqual(
      expect.objectContaining(
        Object.fromEntries(
          Object.entries(expectedRoutes).map(([route, methods]) => [
            route,
            expect.objectContaining(
              Object.fromEntries(
                methods.map((method: string) => [
                  method.toLowerCase(),
                  expect.any(Object),
                ])
              )
            ),
          ])
        )
      )
    );

    // Ensure we have the expected number of routes
    expect(Object.keys(paths)).toHaveLength(Object.keys(expectedRoutes).length);
  });

  test("matches snapshot", () => {
    // Use Vitest's snapshot testing
    expect(output).toMatchSnapshot();
  });
});
