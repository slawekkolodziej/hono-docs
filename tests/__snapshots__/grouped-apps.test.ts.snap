// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`'basepath-grouped' - 'Grouped app with basePath' > matches snapshot 1`] = `
{
  "info": {
    "title": "BasePath Grouped App Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/v1/docs": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "docs": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "docs",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /v1/docs",
        "tags": [
          "BasePath Grouped App",
        ],
      },
    },
    "/v1/docs/{id}": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "docId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "docId",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /v1/docs/{id}",
        "tags": [
          "BasePath Grouped App",
        ],
      },
    },
    "/v1/users": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "users": {
                      "items": {},
                      "type": "array",
                    },
                  },
                  "required": [
                    "users",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /v1/users",
        "tags": [
          "BasePath Grouped App",
        ],
      },
      "post": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "created": {
                      "oneOf": [
                        {},
                        {},
                      ],
                    },
                  },
                  "required": [
                    "created",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated POST /v1/users",
        "tags": [
          "BasePath Grouped App",
        ],
      },
    },
    "/v1/users/{id}": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "userId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "userId",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /v1/users/{id}",
        "tags": [
          "BasePath Grouped App",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "BasePath Grouped App",
    },
  ],
}
`;

exports[`'complex-grouped' - 'Complex grouped app with mixed composâ€¦' > matches snapshot 1`] = `
{
  "info": {
    "title": "Complex Grouped App Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/api/v1/docs": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "docs": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "docs",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/v1/docs",
        "tags": [
          "Complex Grouped App",
        ],
      },
    },
    "/api/v1/docs/{id}": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "docId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "docId",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/v1/docs/{id}",
        "tags": [
          "Complex Grouped App",
        ],
      },
    },
    "/api/v1/users": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "users": {
                      "items": {},
                      "type": "array",
                    },
                  },
                  "required": [
                    "users",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/v1/users",
        "tags": [
          "Complex Grouped App",
        ],
      },
      "post": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "created": {
                      "oneOf": [
                        {},
                        {},
                      ],
                    },
                  },
                  "required": [
                    "created",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated POST /api/v1/users",
        "tags": [
          "Complex Grouped App",
        ],
      },
    },
    "/api/v1/users/{id}": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "userId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "userId",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/v1/users/{id}",
        "tags": [
          "Complex Grouped App",
        ],
      },
    },
    "/auth/login": {
      "post": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "token": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "token",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated POST /auth/login",
        "tags": [
          "Complex Grouped App",
        ],
      },
    },
    "/auth/logout": {
      "post": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "loggedOut": {
                      "oneOf": [
                        {},
                        {},
                      ],
                    },
                  },
                  "required": [
                    "loggedOut",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated POST /auth/logout",
        "tags": [
          "Complex Grouped App",
        ],
      },
    },
    "/health": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "status": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "status",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /health",
        "tags": [
          "Complex Grouped App",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "Complex Grouped App",
    },
  ],
}
`;

exports[`'nested-grouped' - 'Nested grouping' > matches snapshot 1`] = `
{
  "info": {
    "title": "Nested Grouped App Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/api/docs": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "docs": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "docs",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/docs/",
        "tags": [
          "Nested Grouped App",
        ],
      },
    },
    "/api/docs/{id}": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "docId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "docId",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/docs/{id}",
        "tags": [
          "Nested Grouped App",
        ],
      },
    },
    "/api/users": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "users": {
                      "items": {},
                      "type": "array",
                    },
                  },
                  "required": [
                    "users",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/users/",
        "tags": [
          "Nested Grouped App",
        ],
      },
      "post": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "created": {
                      "oneOf": [
                        {},
                        {},
                      ],
                    },
                  },
                  "required": [
                    "created",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated POST /api/users/",
        "tags": [
          "Nested Grouped App",
        ],
      },
    },
    "/api/users/{id}": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "userId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "userId",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /api/users/{id}",
        "tags": [
          "Nested Grouped App",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "Nested Grouped App",
    },
  ],
}
`;

exports[`'simple-grouped' - 'Simple grouped app with .route()' > matches snapshot 1`] = `
{
  "info": {
    "title": "Simple Grouped App Test API",
    "version": "1.0.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/docs": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "docs": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "docs",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /docs",
        "tags": [
          "Simple Grouped App",
        ],
      },
    },
    "/docs/{id}": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "docId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "docId",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /docs/{id}",
        "tags": [
          "Simple Grouped App",
        ],
      },
    },
    "/users": {
      "get": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "users": {
                      "items": {},
                      "type": "array",
                    },
                  },
                  "required": [
                    "users",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /users",
        "tags": [
          "Simple Grouped App",
        ],
      },
      "post": {
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "created": {
                      "oneOf": [
                        {},
                        {},
                      ],
                    },
                  },
                  "required": [
                    "created",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated POST /users",
        "tags": [
          "Simple Grouped App",
        ],
      },
    },
    "/users/{id}": {
      "get": {
        "parameters": [
          {
            "in": "path",
            "name": "id",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "default": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "userId": {
                      "type": "string",
                    },
                  },
                  "required": [
                    "userId",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Default fallback response",
          },
        },
        "summary": "Auto-generated GET /users/{id}",
        "tags": [
          "Simple Grouped App",
        ],
      },
    },
  },
  "tags": [
    {
      "name": "Simple Grouped App",
    },
  ],
}
`;
